network/src/commonMain/kotlin/com/wire/kalium/network/utils/NetworkUtils.kt

```diff
@@ -121,54 +116,62 @@ internal fun <T : Any> NetworkResponse<T>.onFailure(fn: (NetworkResponse.Error)
-                // 400 .. 499
-                when (e.response.status) {
-
-                    // for 401 error the BE return response with content-type: text/html which kalium ktor client
-                    // has no idea how to parse -> app crash
-                    HttpStatusCode.Unauthorized -> {
-                        kaliumLogger.e("Unauthorized request", e)
-                        NetworkResponse.Error(KaliumException.Unauthorized(e.response.status.value))
-                    }
-
-                    else -> try {
-                            e.response.body() as ErrorResponse
-                        } catch (_: NoTransformationFoundException) {
-                            ErrorResponse(e.response.status.value, e.response.status.description, "")
-                        }.let { errorResponse ->
-                            NetworkResponse.Error(kException = KaliumException.InvalidRequestError(errorResponse))
-                        }
-                }
-            }
-            is ServerResponseException -> {
-                // 500 .. 599
-                // TODO: do 500 errors have body
-                NetworkResponse.Error(kException = KaliumException.ServerError(e.response.body()))
+/**
+ * Wraps exceptions thrown during the request or parsing of the response, like
+ * exceptions thrown due to inability to reach the server.
+ * This does **not** handle the response itself (i.e. HTTP status, body, etc.),
+ * but may be used to catch exceptions thrown during those steps.
+ * 
+ * @return [NetworkResponse.Success] if everything went smooth
+ * @return [NetworkResponse.Error] with a [KaliumException.GenericError] in case of an error
+ */
+private inline fun <reified ResponseType: Any> handlingNetworkException(
+    performRequest: () -> NetworkResponse<ResponseType>
+): NetworkResponse<ResponseType> = try {
+    performRequest()
+} catch (e: Exception) {
+    NetworkResponse.Error(KaliumException.GenericError(e))
+}
+
+/**
+ * Wraps a producer of [HttpResponse] and attempts to parse the server response based on the [BodyType].
+ * @return - Successful response (HTTP Status Codes from 200 to 299):
+ * a [NetworkResponse.Success] with the expected [BodyType] will be returned.
+ *
+ * - Unsuccessful response (any other HTTP Status Code):
+ * a [NetworkResponse.Error] with a [KaliumException]. Will try to read it the body as an [ErrorResponse].
+ * It's possible to intercept this and do the mapping to a [NetworkResponse] through [unsuccessfulResponseOverride].
+ *
+ * - Exceptions failure to reach server or parse response:
+ * a [NetworkResponse.Error] containing a [KaliumException.GenericError]
+ *
+ * @param unsuccessfulResponseOverride Allows to intercept any unsuccessful response
+ * and map it to a [NetworkResponse] as needed. Useful when handling custom ErrorBody
+ * @param performRequest the block that will result into the [HttpResponse]
+ * @see KaliumException
+ * @see ErrorResponse
+ */
```

Comment:
Looks like this in IntelliJ:

![grafik](https://user-images.githubusercontent.com/9389043/166800118-5f437741-c03f-40db-a371-52bcf9bc72a7.png)